middle_end/flambda/types/env/typing_env_level.rec.ml

```diff
@@ -259,20 +276,31 @@ let join_types ~env_at_fork envs_with_levels =
             | None, Some use_ty ->
               (* In this case, we haven't yet got a joined type for [name]. *)
               let left_ty =
-                (* If this is the first level in the join, we just need to
-                   make the type suitable for the joined environment, so we
-                   use [Bottom] to avoid losing precision... *)
-                if is_first_join then Type_grammar.bottom_like use_ty
-                (* ...but if this is not the first level in the join, then we
+                (* If this is the first occurrence of [name] in the join,
+                   and [name] was not part of the base environment,
+                   we just need to make the type suitable for the joined
+                   environment, so we use [Bottom] to avoid losing precision...
```

Comment:
If you are motivated, you could add to the comment what situation leads to this to get a better intuition on what should happen.